src/Voron/Impl/Scratch/ScratchBufferPool.cs

```diff
@@ -503,18 +504,29 @@ public void Cleanup()
 
                 // we need to ensure that no access to _recycleArea and _scratchBuffers will take place in the same time
                 // and only methods that access this are used within write transaction
+
+                // additionally we must not allow any transaction (even read one) to start because it uses GetPagerStatesOfAllScratches() which
+                // returns _pagerStatesAllScratchesCache  we're updating here
+
+                IDisposable exitPreventNewTransactions = null;
+
                 try
                 {
-                    using (_env.WriteTransaction(context: byteStringContext))
+                    if (_env.TryPreventNewReadTransactions(TimeSpan.FromSeconds(30), out exitPreventNewTransactions))
```

Comment:
Doesn't this now expose us to a _write_ transaction doing this?

Reply:
No. The method is actually misleading - it prevents from creating _any_ new transaction. I have renamed it and force pushed changes